lib/python/compliance_checker_test.py

```diff
@@ -0,0 +1,135 @@
+#!/usr/bin/env python
+"""
+Script to facilitate the test of the compliance checker across a selection of IMOS/AODN NetCDF files. This helps making
+sure the upgrade to a new version of the imos plugin/ioos compliance checker doesn't make files previously passing the
+checker not passing it anymore.
+
+The different checker suites tests are
+cf:1.6
+imos:1.4
+imos:1.3
+ghrsst:latest
+
+The script reads a json config file [CONFIG_FILE]. This config file lists for each NetCDF to test:
+ - The path of the NetCDF on the imos s3 bucket
+ - checks which should pass (checks_success)
+ - criteria to apply (normal, lenient ... default is normal)
+ - checks expected to fail (checks_fail)
+ - checks to skip (skip_checks" such as "check_convention_globals")
+
+Each file will be downloaded in a temporary folder. The compliance checker will be run on each of those NetCDF.
+
+The results will be output to a temporary directory (see outfile_path variable):
+* a json file with a similar structure to the CONFIG_FILE containing the different results
+* if some tests fail, the compliance checker output files will be saved in the same output directory
+
+If all test succeeded, the string 'False' should not exist in the output json file. If "False"
+is to be found, this means a test didn't output the required result.
+
+author: Besnard, Laurent
+"""
+import json
+import os
+import tempfile
+
+import cf_units
+import compliance_checker
+from six.moves.urllib.request import urlretrieve
+
+from util import pass_netcdf_checker
+
+TEST_ROOT = os.path.join(os.path.dirname(__file__))
+CONFIG_FILE = os.path.join(TEST_ROOT, "compliance_checker_imos_files_config.json")
+OUTPUT_DIR = tempfile.mkdtemp(prefix='compliance_checker_testing_results_')
+
+with open(CONFIG_FILE, 'r') as f:
+    compliance_config = json.load(f)
+
+print("compliance checker {cc_version}\ncf units {cf_version}".format(cc_version=compliance_checker.__version__,
+                                                                      cf_version=cf_units.__version__))
+
+# collection is equivalent to a facility/sub-facility in the json file
+for collection in compliance_config:
```

Comment:
I know it's a script but there is too much done here. I would simplify it in the following way:

1. Function to create a sub_collections list of dicts.
2. the dicts would be a single-level/flat dictionary with the specify options (file_url,tempfile_obj,...). 
3. Allow a "collection first, execute later" design by lazy evaluation of the calls to pass_netcdf_checker and urlretrieve. For example, it's easy to build a list of all the pass/fail tests to be run before actually running them. As well, urlretrieve could also be lazy so we only downlaod when the first test is run.  This also allow gathering all information before actually running the tests - easy to debug by just inspecting the dict keys.
4. Have a function to evaluate the lazy functions and handle the exceptions/try/cases.

Reply:
>  there is too much done here

I agree. will need to look into it again